src/app/Dashboard/dashboard-utils.tsx

```diff
@@ -0,0 +1,170 @@
+ * derivative works of, display, perform, and distribute the Software and make,
+ * use, sell, offer for sale, import, export, have made, and have sold the
+ * Software and the Larger Work(s), and to sublicense the foregoing rights on
+ * either these or other terms.
+ *
+ * This license is subject to the following condition:
+ * The above copyright notice and either this complete permission notice or at
+ * a minimum a reference to the UPL must be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+import cryostatLogo from '@app/assets/cryostat_icon_rgb_default.svg';
+import { gridSpans } from '@patternfly/react-core';
+import { FileIcon, UserIcon } from '@patternfly/react-icons';
+import { nanoid } from '@reduxjs/toolkit';
+import React from 'react';
+
+export const DEFAULT_DASHBOARD_NAME = 'Default';
+export const DRAGGABLE_REF_KLAZZ = `draggable-ref`;
+export const LAYOUT_TEMPLATE_DESCRIPTION_WORD_LIMIT = 100;
+
+export const DashboardLayoutNamePattern = /^[a-zA-Z0-9_.-]+( [a-zA-Z0-9_.-]+)*$/;
+export const LayoutTemplateDescriptionPattern = /^[a-zA-Z0-9\s.,\-'";?!@#$%^&*()[\]_+=:{}]*$/;
+
+export interface CardConfig {
+  id: string;
+  name: string;
+  span: gridSpans;
+  props: object;
+}
+
+export type SerialCardConfig = Omit<CardConfig, 'id'>;
+
+export const mockSerialCardConfig: SerialCardConfig = {
+  name: 'Default',
+  span: 12,
+  props: {},
+};
+export interface DashboardLayout {
+  name: string;
+  cards: CardConfig[];
+  favorite: boolean;
+}
+
+export type SerialDashboardLayout = Omit<DashboardLayout, 'name' | 'cards'> & { cards: SerialCardConfig[] };
+
+// only name and vendor are needed to identify a template
+export type LayoutTemplateRecord = Pick<LayoutTemplate, 'name' | 'vendor'>;
+
+export enum LayoutTemplateVersion {
+  'v2.3' = 'v2.3',
+}
+
+export enum LayoutTemplateVendor {
+  BLANK = 'Blank',
+  CRYOSTAT = 'Cryostat',
+  USER = 'User-supplied',
+}
+
+export interface LayoutTemplate {
+  name: string;
+  description: string;
+  cards: SerialCardConfig[];
+  version: LayoutTemplateVersion;
+  vendor: LayoutTemplateVendor;
+}
+
+export type SerialLayoutTemplate = Omit<LayoutTemplate, 'vendor'>;
+
+export const mockSerialLayoutTemplate: SerialLayoutTemplate = {
+  name: 'Default',
+  description: 'Default.',
+  cards: [] as SerialCardConfig[],
+  version: LayoutTemplateVersion['v2.3'],
+};
+
+export interface LayoutTemplateProviderProps {
```

Comment:
I think its the type of the context value. Lets call it something else.

Reply:
Maybe name it `LayoutTemplateController` or something else that means controlling?

Reply:
What about just `LayoutTemplateContextProps`?

Reply:
its not exactly a props tho so...no `props`?